### ISSUE TITLE
FFmpegInterop really struggles to play some 1080p videos and all 4k videos

### ISSUE THREAD

**Issue by viswanathr88** [Microsoft/FFmpegInterop#202]
iPhones now captures videos in 1080p and 4k and as a result they have become so common. I have my own uwp video project where I'm trying to play such videos. I had lot of issues with framerate, that's why I checked this project. 
1. First of all, audio and video pipeline are not split up into separate threads in ffmpeg interop. That way, cores can be better utilized. Also, decoding and scaling the frame too. This made 1080p videos play much smoother, but 4k was still pretty poor.
2. Then I able to solve most of the frame rate issues by asking FFMPEG to multi-thread the decode process (which was also recently added by FFmpegInterop). But the problem now seems to be sws_scale. It takes ~30ms on my machine (4 core 8 thread - i7) to scale one frame.
Note: I'm only talking about the case where Ffmpeg does the decoding and not Windows.

My question is: Is there any way by which sws_scale can be avoided all together and somehow we could render the YUV frame onto the screen directly? I feel that's the only way we could keep up the frame rate. VLC Store app is somehow able to do a fantastic job, even using less CPU.

**Comment by mcosmin222** [comment]
Have you tried the latest verion of this repository? We have multithreaded video decoding now.

About VLC, my guess is that they use hardware acceleration for that, which this library can also do.

**Comment by viswanathr88** [comment]
I posted this only after trying it out with the latest version of ffmpeginterop.
1. Pick a 4k video recorded in an iPhone or android phone
2. Start ffmpeginterop (c#) and set Force Decode Video and Force Decode Audio to be true.
3. Play the 4k video.

Ffmpeg can play the video for 2 seconds and then buffers for the next 4 seconds, then plays 2 seconds and buffers again and so on.

**Comment by mcosmin222** [comment]
The only way to render the YUV frame directly is to use directx. This is sadly not the purpose of this library so far, we are just tieing things into a consumer MediaPlayer(Element).

VLC does something similar with Video decoding. But VLC on the other hand has its own custom media players.

**Comment by reego-fr** [comment]
Hi, I am surprised sws_scale can be so slow on a desktop but if the following still applies it could be replaced by a faster implementation if it is possible for you to go GPL. [This link](http://www.ffmpeg-archive.org/sws-scale-slow-with-big-resolutions-td2272531.html) contains a suggestion to compile ffmpeg with --enable-gpl. 
The best of course would be to get rid of it when possible, but I am not sure what pixel formats the current platforms are limited to.

**Comment by lukasf** [comment]
Wow 33ms, really? That's extremely poor performance. I was also surprised to see that HEVC 4k videos still stutter with multi threaded decoding, but the CPU is only at 30%. Looks like you found the bottleneck!

Unfortunately, it does not seem to be possible to get MediaElement to play the native YUV420P format. I just tried all available YUV subtypes and none worked. DirectShow lists some more formats but they do not seem to work in MediaFoundation (I tried to set the subtype to Guid of IMC2/IMC4, but it is not accepted). So we definitely need to do a conversion. Has someone already tried with a GPL build?

**Comment by viswanathr88** [comment]
Actually, I have some more info on this. I was actually able to play the YUV420P format directly using MediaEncodingSubTypes::Iyuv (seems to be the same thing). This avoided the scaling, but the video would stutter even worse in that case, which was very bizarre. Then I read some msdn article on encoding that said NV12 is a much more efficient format for rendering. YUV420 is more efficient for transcoding, not so much for playing. So, playing YUV420P directly made it worse, at least with the MediaElement. So we have to convert it to NV12. libavc is also not capable of outputting video directly in Nv12, it can only output to YUV420p. 

Also it is sws_scale + copying the pixels onto another buffer that led to slow performance (~30 ms). Each frame is 10KB and memcopying this also seems to take a variable amount of time depending on how much of CPU is available to the thread.

I do agree that decoding and rendering on hardware directly is the best way for this. But with a machine as powerful as the one I have (12gigs ram + 4 core 8 thread i7 3.0Ghz), we should be able to render this. CPU usage stays at 40-45% right now. Wondering how to better utilize it.

**Comment by lukasf** [comment]
Really, Iyuv works for you? When I try Iyuv, I get green tint all over the image. But playback is a lot better for me. Still not sufficient, but I'd say about twice as fast. Did you do anything special to get it working?

I also tried RGB, but performance did not improve compared to NV12 (plus, image was upside down).

I wonder if it would be possible to decode directly into a target buffer, instead of ffmpeg creating a buffer, decoding into it, and us copying the stuff out again and into our internal buffer... If we do scaling, then scaling into a new buffer is not an issue. But if we can output the image directly, I'm pretty sure zero-copy would help here as well.

**Comment by viswanathr88** [comment]
With Iyuv, you just have to make sure you are copying the planes properly. If you are seeing a green tint, it's possible you are completely missing the U or V channel, presumably making it faster.
With YUV420p, you are going to have data[0] representing Y data, data[1] holding U data and data[2] holding V data. And for every 4 Y samples, there should be 1 U sample and 1 V sample making it 4*8 + 8 + 8 = 48. 48/4 = 12 bits per pixel. For Y, you have to memcpy everything, for U and V you will have to copy linesize / 2 into the array. Then passing this to the MediaStreamSample will play with Iyuv.

**Comment by reego-fr** [comment]
I could display something with MediaEncodingSubTypes::Yv12 too and display a 4:2:0 source without the need of sws_scale. 
I did not see any improvements when it was disabled. It looks like the array assignment and WriteBytes are the most time consuming calls, at least in my case. Maybe sws_scale can be slower when doing other type of conversion or scaling but yuv420p -> nv12 does not look that bad, with only ~3ms poorly measured penalty from what I saw.

**Comment by lukasf** [comment]
Thanks for the hints, I can now render Iyuv correctly. With my changes, I am able to decode 50Mbit 4K HEVC files from my DSLR without any interruptions! CPU is at 75% (Core i7 8 threads @ 2,8GHz). Are you sure you tested a release build and without debugger attached? Otherwise, I can create a branch if you like, so you can test my solution. Still need to cleanup the code a bit.

With sws_scaler I had regular breaks with CPU at ~45%. So direct rendering in Iyuv improved performance by more than 60%! 

@reego-fr Did you test with 4K files? I did not see problems with normal files, but with 4K, the scaler is obviously a problem.

Still we have unneccessary copy operations on each frame, due to use of DataWriter. It would be a lot better if we allocate buffers ourselfes (with correct continuous target memory layout) and have ffmpeg (or the scaler) render into them directly. Then we can return that buffer directly as IBuffer result and have zero copy. Combine that with preallocated, cached buffers (to avoid malloc/free on every frame) and we should be very performant.

**Comment by reego-fr** [comment]
Yes I tried with a 4k h264 file, but to be more precise while I could only see minor differences when switching between scaler output pixel format (nv12 <-> yuv420p), copy operations always looked slower in my tests, debugger attached though. 
Still, every millisecond is precious and it will be better to avoid layout conversions if the media player can natively display the decoder output pixel format. 
On copy operations side, I experimented with ArrayReference which apparently was faster that using the array contructor, it is to be confirmed as I don't have a good test protocol.

**Comment by lukasf** [comment]
The intention is of course to not use the scaler at all. If you just change the output, scaler still needs to copy the frame. My tests were done without any scaler calls. I have a pretty good solution now I think, will publish a branch very soon.

**Comment by lukasf** [comment]
Branch is there: lukasf/yuv-output

I will also do some testing with ArrayReference

**Comment by lukasf** [comment]
Using ArrayReference further improves performance by eliminating one unneccessary copy operation. This improved performance on the same 4K file, playing now at ~65% CPU.

One interesting side note: After removing sws_scaler, I now see GPU usage in task manager, which was not the case before. Looks like MF will do the scaling to target format in hardware now. Nice! ;)

PR created: #204

**Comment by reego-fr** [comment]
I'm happy to hear that and to see you switched to ArrayReference everywhere. With the optimisation applied to the audio samples too my sample file now plays smoothly, far better than what I could get only focusing on video frames.

**Comment by touseefbsb** [comment]
@reego-fr  @viswanathr88  any update ? is this problem fixed? or still you guyx having the same problem?

**Comment by reego-fr** [comment]
It is obviously a lot better now, with no more pixel format conversion when the decoder output is yuv420, which should be the case for most high resolution media out there, and an additional copy operation removed in all cases. 
I don't know if it is possible to go further as @lukasf mentioned and skip the last copy operation, but I think right know we are very close of having the most efficient pipeline it is possible to get.

**Comment by mcosmin222** [comment]
So then perhaps it is time for the merge?

**Comment by reego-fr** [comment]
Yes, I think it is good to merge. 
It's weird, on mobile I can see the merge button... but I don't know if I should try clicking it :-) On desktop I see a greyed merge button, I don't think I am allowed to merge.

**Comment by mcosmin222** [comment]
Same here. But you cant actually merge anyway.

**Comment by touseefbsb** [comment]
you will give a pull request I guess and anyone of the admins of the project will merge it @reego-fr

**Comment by khouzam** [comment]
That is awesome guys. I've been on vacation for the last few weeks but will take a look and merge this soon.

**Comment by mcosmin222** [comment]
Perhaps we could merge all 4 PRs.

**Comment by touseefbsb** [comment]
yeah our aim should be to make this library ALL IN ONE for media apps of uwp. great job with this contribution guys :)

**Comment by viswanathr88** [comment]
This is a great improvement. But I feel rendering directly in IYUV is working only because you are running a machine with a GPU and MF is using hardware to render it directly. On my machine I only have Intel HD integrated graphics. When I use IYUV, I'm getting lesser framerate that when scaling to NV12 and running it. Is there anyone else who could try this and ensure this is a one off thing?

**Comment by mcosmin222** [comment]
which intel HD card?

**Comment by viswanathr88** [comment]
https://www.intel.com/content/www/us/en/support/products/81500/graphics-drivers/graphics-for-2nd-generation-intel-processors/intel-hd-graphics-3000.html.

It's about 6 years old. It's integrated on the motherboard, it's not dedicated.

**Comment by mcosmin222** [comment]
Hmmm.
Good point, we haven't thought of that.

Perhaps there is a way to query the hardware acceleration?

**Comment by touseefbsb** [comment]
how does the vlc libraries do the hardware accelration , cnt tht be achieved with ffmpeg as well?

**Comment by mcosmin222** [comment]
vlc doesn't support hardware acceleration for all formats.

**Comment by lukasf** [comment]
@viswanathr88 While my laptop does have a dedicated GPU, playback happens on the iGPU. But it is a Core i7 7th generation, so it's probably a lot more powerful, and it is already at ~30% when playing 4K files. Since your iGPU is a lot slower, it might just not be able to convert the format of 4K frames in real time. I don't think that MF uses "real" hardware acceleration here. I'd guess it is just a shader running on the GPU to do the format conversion. In that case it could be hard to find out which GPU is fast enough and which is not.

If you have Windows 10 Fall Creators Update, can you check your GPU utilization in task manager during playback? That would tell us if GPU does the format conversion or not. With software scaler, GPU should be at 0%.

Anyways it's probably best if we make this configurable (both for yuv and argb output). Maybe now would be the right time to introduce a configuration class to pass into constructor? I really wouldn't like to add two more parameters to the already long list (we already had a discussion on that).

**Comment by mcosmin222** [comment]
I think we can query if a video card supports directx11 or 12. HEVC should be decodable on GPU on something which supports DX11

**Comment by viswanathr88** [comment]
@lukasf I do have Fall Creators Update. I will let you know when I head home and run the app, on whether GPU is being utilized or not.

**Comment by lukasf** [comment]
@mcosmin222 This is all about software decoding. It is only the question if the format conversion from (already uncompressed) YUV->NV12 frames is done by ffmpeg sws_scaler or done by MF (probably on the GPU). Decoding happens in software by ffmpeg any way. HEVC hardware decoders have only been added in 7th generation Core CPUs, no way this can be found in a 6 year old PC.

**Comment by lukasf** [comment]
@viswanathr88 One other thing you can try as well is running the file on my hevc-passthrough branch. Even if your iGPU cannot hardware decode, maybe the MF pipeline will do a better job for your specific hardware configuration.

**Comment by mcosmin222** [comment]
@lukasf , my point was about detecting which video card is capable of decoding it, and when to use sw_scale instead. Of course, we could configure it, but I think auto detection should also work. I will study the problem further.

**Comment by lukasf** [comment]
@mcosmin222 By "decoding" you mean the format conversion from YUV to NV12? I would not call that decoding, because the YUV image is already a decoded image. FFmpeg has already done the HEVC decoding in software, and the YUV image is the output of the HEVC software decoder. We are just converting the YUV decoder output to a different representation (NV12) that is easier to render for the GPU. That is not decoding, it is a pixel format conversion, and it does not have much to do with HEVC. Most video decoders will output YUV420 frames.

If the GPU can hardware decode HEVC, then my hevc-passthrough branch solution should be used, it will be blazing fast. But here the question is, if the GPU can NOT hardware decode HEVC (and we use ffmpeg software decoder), can the GPU do at least the YUV->NV12 format conversion (fast enough) or shall we also do the format conversion in software with sws_scale. Querying HEVC hardware support should be easy. But querying YUV->NV12 performance is a different thing. I don't know if this can be done in a reliable way. But if you find a way, that would be great of course.

Sorry if I sound picky. I just want to clarify what we are talking about, to avoid confusion.

**Comment by mcosmin222** [comment]
@lukasf I am referring to using MF to do the job of sw_scale on the GPU.

**Comment by lukasf** [comment]
@mcosmin222 Alright, now we are clear. It will be interesting to hear what @viswanathr88 will see in task manager when MF does the format conversion.

**Comment by viswanathr88** [comment]
@lukasf I tried to sync your branch and run a few videos I have. Unfortunately, if I select Force Decode Video, it actually fails with an error code. If I uncheck Force Decode, it works fine (it goes through MP). Not sure if this is some local issue or something widespread. Are there some logs I can capture to help figure this out?

**Comment by lukasf** [comment]
@viswanathr88 Are you sure you run the code exactly as checked in on that branch? Please do not try to mix this with other changes you might have, and use one of the provided sample apps. If in doubt, clone the git repo again into a new folder, and run the app from there.

If you are running like that: The error code could help, but also please try to put breakpoints in WriteAVPacketToStream and GetNextSample of UncompressedVideoSampleProvider and step through, to see if an error occurs inside our code. Currently there is no logging, which makes it hard to diagnose issued only occuring one a specific machine.

**Comment by MouriNaruto** [comment]
I have tested the @lukasf 's work in three days ago. But find its performance same as the master branch's. So I have to try to optimize it by myself and I found something works.

First, the BufferTime property of MediaStreamSource should set to 0 to disable buffering.
It can extremely significant reduce the memory usage. (From 10GB+ to 700MB, in my 4k 120fps avc 8 bit video.)

Second, I use the performance profiler in visual studio and find the DataWriter object in the UncompressedVideoSampleProvider class eats 10% of my cpu usage. So I have written a class to convert pointer to IBuffer directly. 

```
class BufferReference : public RuntimeClass<
	RuntimeClassFlags<RuntimeClassType::WinRtClassicComMix>,
	abi_IBuffer,
	abi_IBufferByteAccess>
{
private:
	UINT32 m_Capacity;
	UINT32 m_Length;
	byte* m_Pointer;

public:
	virtual ~BufferReference() throw()
	{
	}

	STDMETHODIMP RuntimeClassInitialize(
		byte* Pointer, UINT32 Capacity) throw()
	{
		m_Capacity = Capacity;
		m_Length = Capacity;
		m_Pointer = Pointer;
		return S_OK;
	}

	// IBufferByteAccess::Buffer
	STDMETHODIMP Buffer(byte **value) throw()
	{
		*value = m_Pointer;
		return S_OK;
	}

	// IBuffer::get_Capacity
	STDMETHODIMP get_Capacity(UINT32 *value) throw()
	{
		*value = m_Capacity;
		return S_OK;
	}

	// IBuffer::get_Length
	STDMETHODIMP get_Length(UINT32 *value) throw()
	{
		*value = m_Length;
		return S_OK;
	}

	// IBuffer::put_Length
	STDMETHODIMP put_Length(UINT32 value) throw()
	{
		if (value > m_Capacity)
			return E_INVALIDARG;
		m_Length = value;
		return S_OK;
	}
};

// The M2MakeIBuffer function retrieves the IBuffer object from the provided 
// raw pointer.
//
// Parameters:
//
// Pointer
//     The raw pointer you want to retrieve the IBuffer object.
// Capacity
//     The size of raw pointer you want to retrieve the IBuffer object.
//
// Return value:
//
// If the function succeeds, the return value is the IBuffer object from the 
// provided raw pointer. If the function fails, the return value is nullptr.
//
// Warning: 
// The lifetime of the returned IBuffer object is controlled by the lifetime of
// the raw pointer that's passed to this method. When the raw pointer has been 
// released, the IBuffer object becomes invalid and must not be used.
IBuffer^ M2MakeIBuffer(byte* Pointer, UINT32 Capacity) throw()
{
	IBuffer^ buffer = nullptr;
	
	ComPtr<BufferReference> bufferReference;
	if (SUCCEEDED(MakeAndInitialize<BufferReference>(
		&bufferReference, Pointer, Capacity)))
	{
		buffer = reinterpret_cast<IBuffer^>(bufferReference.Get());
	}

	return buffer;
}
```
Then I found the maximum fps from 70 to 100. (The target is 120, just like PotPlayer on my machine.) 



I hope it can help you.

Mouri
